### PR TITLE
[url_launcher] Convert Android to Pigeon

### PR DESCRIPTION
**Pull request by stuartmorgan**
Converts from direct use of method channels to using Pigeon.

Adds unit testing of the actual implementation logic. The previous unit tests only covered the intermediate class that converted method calls/arguments into Java calls, which has been removed since that's now handled by Pigeon.

Part of https://github.com/flutter/flutter/issues/117915

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java [hunk 1]
```diff
@@ -15,99 +15,122 @@
...
+    this.applicationContext = context;
+    this.intentResolver = intentResolver;
+  }
+
+  UrlLauncher(@NonNull Context context) {
+    this(
+        context,
+        intent -> {
+          ComponentName componentName = intent.resolveActivity(context.getPackageManager());
+          return componentName == null ? null : componentName.toShortString();
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java:53]: I was able to narrow the code that's *not* being unit tested from "all of `UrlLauncher`" to just these two lines (which come from the old `canLaunch`). I spent some time looking at this, and didn't see how to mock enough of `context` to make `intent.resolveActivity` return a dummy `ComponentName`, so I just extracted this one small piece into something tests could replace.

File: packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java [hunk 2]
```diff
@@ -15,99 +15,122 @@
...
 
-  /** Result of a {@link #launch} call. */
-  enum LaunchStatus {
-    /** The intent was well formed. */
-    OK,
-    /** No activity was found to launch. */
-    NO_ACTIVITY,
-    /** No Activity found that can handle given intent. */
-    ACTIVITY_NOT_FOUND,
+  private static @NonNull Bundle extractBundle(Map<String, String> headersMap) {
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java:121]: This, and the strings in the error in `ensureActivity`, are the surviving code from `MethodCallHandlerImpl`.

**Review by gmackall — APPROVED** [review]
This LGTM, though I haven't reviewed any conversions to pigeon before. In your opinion is this a change that it would be good for me to find a second reviewer on, or is the perceived risk low?

**Comment by stuartmorgan** [comment]
The pigeon surface here is quite small, so between that and the expanded native unit testing I think the risk is pretty low.

**Comment by bparrishMines** [comment]
@stuartmorgan This is causing a failure on main:

```
Analyzing url_launcher_android...

   info - test/url_launcher_android_test.dart:89:7 - Missing an 'await' for the 'Future' computed by this expression. Try adding an 'await' or wrapping the expression un 'unawaited'. - unawaited_futures
   info - test/url_launcher_android_test.dart:104:7 - Missing an 'await' for the 'Future' computed by this expression. Try adding an 'await' or wrapping the expression un 'unawaited'. - unawaited_futures
   info - test/url_launcher_android_test.dart:170:7 - Missing an 'await' for the 'Future' computed by this expression. Try adding an 'await' or wrapping the expression un 'unawaited'. - unawaited_futures
   info - test/url_launcher_android_test.dart:185:7 - Missing an 'await' for the 'Future' computed by this expression. Try adding an 'await' or wrapping the expression un 'unawaited'. - unawaited_futures
```

**Comment by bparrishMines** [comment]
https://github.com/flutter/packages/pull/4086

**Comment by rafaelmeteoro** [comment]
I catch an error on build android:
```
/opt/flutter_3.3.0/.pub-cache/hosted/pub.dartlang.org/url_launcher_android-6.0.35/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java:66: error: cannot find symbol01:34
    if (BuildConfig.DEBUG) {
        ^
  symbol:   variable BuildConfig01:34
  location: class UrlLauncher
  ```
  anyone else with this?

**Comment by bparrishMines** [comment]
@rafaelmeteoro Could you open a new issue with an explanation of when you are seeing this error? https://github.com/flutter/flutter/issues/new/choose

**Pull request merged**